# Think Python

## Chapter 15 - Classes and objects

### 15.1 Programmer-defined types

*HTML of this chapter in "Think Python 2e" can be found [here](http://greenteapress.com/thinkpython2/html/thinkpython2016.html "Chapter 15").*

*__Creating a new `class`:__*

In [1]:
class Point:
    """Represents a point in 2-D space."""

*As an exercise, write a function called `distance_between_points` that takes two Points as arguments and returns the distance between them.*

In [2]:
import math

def distance_between_points(a, b):
    return math.sqrt((b.x - a.x)**2 + (b.y - a.y)**2)

In [3]:
a = Point()
a.x, a.y = 4, 4

b = Point()
b.x, b.y = 0, 0

distance_between_points(a, b)

5.656854249492381

### 15.2 Attributes



In [4]:
def print_point(p):
    print('(%g, %g)' % (p.x, p.y))

### 15.3 Rectangles

In [5]:
class Rectangle:
    """
    Represents a rectangle.
    
    attributes: width, height, corner
    """

In [6]:
box = Rectangle()
box.width = 100.0
box.height = 200.0
box.corner = Point()
box.corner.x = 0.0
box.corner.y = 0.0

### 15.4 Instances as return values

In [7]:
def find_center(rect):
    p = Point()
    p.x = rect.corner.x + rect.width / 2
    p.y = rect.corner.y + rect.height / 2
    return p

### 15.5 Objects are mutable



*As an exercise, write a function named `move_rectangle` that takes a Rectangle and two numbers named `dx` and `dy`. It should change the location of the rectangle by adding `dx` to the `x` coordinate of corner and adding `dy` to the `y` coordinate of corner.*

In [8]:
def move_rectangle(rect, dx, dy):
    rect.corner.x += dx
    rect.corner.y += dy
    return rect

In [9]:
move_rectangle(box, 100, 100)
center = find_center(box)
print_point(center)

(150, 200)


### 15.6 Copying




*As an exercise, write a function named `move_rectangle` that takes a Rectangle and two numbers named `dx` and `dy`. It should change the location of the rectangle by adding `dx` to the `x` coordinate of corner and adding `dy` to the `y` coordinate of corner.*

In [10]:
def move_rectangle(rect, dx, dy):
    rect.corner.x += dx
    rect.corner.y += dy
    return rect

*As an exercise, write a version of `move_rectangle` that creates and returns a new Rectangle instead of modifying the old one.*

In [11]:
import copy

def move_new_rectangle(rect, dx, dy):
    
    new_rect = copy.deepcopy(rect)
    new_rect.corner.x += dx
    new_rect.corner.y += dy
    return new_rect

### 15.9  Exercises

#### Exercise 1  

*Write a definition for a class named `Circle` with attributes `center` and `radius`, where `center` is a `Point` object and radius is a number.*

*Instantiate a `Circle` object that represents a circle with its center at (150, 100) and radius 75.*



In [12]:
class Circle:
    """
    Represents a circle.
    
    Attributes: radius, center
    """

In [13]:
roundy = Circle()
roundy.radius = 75
roundy.center = Point()
roundy.center.x = 150.0
roundy.center.y = 100.0

*Write a function named `point_in_circle` that takes a `Circle` and a `Point` and returns `True` if the `Point` lies in or on the boundary of the circle.*



In [14]:
def point_in_circle(circle, point):
    """
    Returns True if Point point lies within or on the 
    boundary of Circle circle.
    """
    return distance_between_points(circle.center, point) <= circle.radius

In [15]:
# point on the boundary

test_point_a = Point()
test_point_a.x, test_point_a.y = 225.0, 100.0

point_in_circle(roundy, test_point_a)

True

In [16]:
# point within the boundary

test_point_b = Point()
test_point_b.x, test_point_b.y = 150.0, 160.0

point_in_circle(roundy, test_point_b)

True

In [17]:
# point outside the boundary

test_point_c = Point()
test_point_c.x, test_point_c.y= 225.0, 175.0

point_in_circle(roundy, test_point_c)

False

*Write a function named `rect_in_circle` that takes a `Circle` and a `Rectangle` and returns `True` if the `Rectangle` lies entirely in or on the boundary of the circle.*



In [18]:
def find_rect_corners(rect):
    """
    Takes Rectangle rect and returns a list with four 
    Point objects representing the 
    coordinate of the corners of rect.
    """
    
    bl = rect.corner
    tl, tr, br = Point(), Point(), Point()
    tl.x, tl.y = rect.corner.x, rect.corner.y + rect.height
    tr.x, tr.y = rect.corner.x + rect.width, rect.corner.y + rect.height
    br.x, br.y = rect.corner.x + rect.width, rect.corner.y
    
    return [bl, tl, tr, br]
    

In [19]:
def rect_in_circle(circle, rect):
    """
    Returns True if Rectangle rect lies
    wholly within Circle circle.
    """
    
    within_circle = []
    for rc in find_rect_corners(rect):
        # will append True or False to within_circle for each corner
        within_circle.append(point_in_circle(circle, rc))
        
    # Since True values are represented by the value 1,
    # if all four corners are within the circle the sum
    # of the values in within_circle will be 4.
    
    return sum(within_circle) == 4

In [20]:
# small_box should easily fit in the circle

small_box = Rectangle()
small_box.width = 50.0
small_box.height = 60.0
small_box.corner = Point()
small_box.corner.x = 125.0
small_box.corner.y = 70.0

In [21]:
rect_in_circle(roundy, small_box)

True

In [22]:
# half of big_box circle will be in the circle, half out

big_box = Rectangle()
big_box.width = 75.0
big_box.height = 75.0
big_box.corner = Point()
big_box.corner.x = 75.0
big_box.corner.y = 50.0

In [23]:
rect_in_circle(roundy, big_box)

False

*Write a function named `rect_circle_overlap` that takes a `Circle` and a `Rectangle` and returns `True` if any of the corners of the `Rectangle` fall inside the circle. Or as a more challenging version, return `True` if any part of the `Rectangle` falls inside the circle.*

In [24]:
def rect_circle_overlap(circle, rect):
    """
    Returns True if any corner of Rectangle rect
    lies within Circle circle.
    """
    
    within_circle = []
    for rc in find_rect_corners(rect):
        # will append True or False to within_circle for each corner
        within_circle.append(point_in_circle(circle, rc))
        
    # Since True values are represented by the value 1,
    # if at least one corner is within the circle, the sum
    # of the values in within_circle will be one or more.
    
    return sum(within_circle) >= 1

In [25]:
# overlap_rect partially overlaps the circle

overlap_rect = Rectangle()
overlap_rect.width = 100.0
overlap_rect.height = 100.0
overlap_rect.corner = Point()
overlap_rect.corner.x = 180.0
overlap_rect.corner.y = 110.0

In [26]:
rect_circle_overlap(roundy, overlap_rect)

True

In [27]:
# nonoverlap_rect and circle have no overlapping parts

nonoverlap_rect = Rectangle()
nonoverlap_rect.width = 100.0
nonoverlap_rect.height = 100.0
nonoverlap_rect.corner = Point()
nonoverlap_rect.corner.x = 280.0
nonoverlap_rect.corner.y = 110.0

In [28]:
rect_circle_overlap(roundy, nonoverlap_rect)

False

*__The more advanced version: if any part of the rectangle is within the circle, then at least one point of the border of the rectangle will fall within the circle.  So we will create lists of all of the points on the four sides of the rectangle, and iteratively check if any of these points lie within the circle.__*

In [29]:
def adv_rect_circle_overlap(circle, rect):
    """
    Returns True if any part of Rectangle rect lies 
    within Circle circle
    """
    
    # lists to hold points respectively for the bottom, top, 
    # left, and right borders of the rectangle
    
    bb, tb, lb, rb = [], [], [], []
    
    # finding all the points for the bottom border:
    
    for i in range(int(rect.width) + 1):
        newbp = Point()
        newbp.x = rect.corner.x + i
        newbp.y = rect.corner.y
        bb.append(newbp)
    
    # finding all the points for the top border:
    
    for i in range(int(rect.width) + 1):
        newtp = Point()
        newtp.x = rect.corner.x + i
        newtp.y = rect.corner.y + rect.height
        tb.append(newtp)
    
    # finding all the points for the left border:
    
    for i in range(int(rect.height) + 1):
        newlp = Point()
        newlp.x = rect.corner.x 
        newlp.y = rect.corner.y + i
        lb.append(newlp)
        
    # finding all the points for the right border:
    
    for i in range(int(rect.height) + 1):
        newrp = Point()
        newrp.x = rect.corner.x + rect.width
        newrp.y = rect.corner.y + i
        rb.append(newrp)
        
    # checking all the points
    
    allb = [bb, tb, lb, rb]
    overlap_points = []
    for bo in allb:
        for b in bo:
            overlap_points.append(point_in_circle(circle, b))
    
    # Since True values are represented by the value 1,
    # if at least one point is within the circle, the sum
    # of the values in within_circle will be one or more.   
    
    return sum(overlap_points) >= 1

In [30]:
# slight_overlap should slightly overlap

slight_overlap = Rectangle()
slight_overlap.width = 76.0
slight_overlap.height = 200.0
slight_overlap.corner = Point()
slight_overlap.corner.x = 0.0
slight_overlap.corner.y = 0.0

In [31]:
adv_rect_circle_overlap(roundy, slight_overlap)

True

In [32]:
# outside_rect should be entirely outside the circle

outside_rect = Rectangle()
outside_rect.width = 74.0
outside_rect.height = 200.0
outside_rect.corner = Point()
outside_rect.corner.x = 0.0
outside_rect.corner.y = 0.0

In [33]:
adv_rect_circle_overlap(roundy, outside_rect)

False

#### Exercise 2  

*Write a function called `draw_rect` that takes a `Turtle` object and a `Rectangle` and uses the `Turtle` to draw the `Rectangle`. See Chapter 4 for examples using Turtle objects.*

*__As was the case with the exercises in Chapter 4, I can't use the `Turtle` library inside jupyter notebooks, so I wrote the code with VS Code and executed it from the command line.  Below is this code which can be found in my GitHub repo [here](https://github.com/Sturzgefahr/ThinkPython/blob/master/Think%20Python%20-%20Chapter%2015/draw_rect.py "draw_rect.py").__*

```
import turtle
bob = turtle.Turtle()

class Point:
    """Represents a point in 2-D space."""

class Rectangle:
    """
    Represents a rectangle.
    
    attributes: width, height, corner
    """

    
def draw_rect(t, rect, color = "black"):
    """
    Takes a Rectangle object and a Turtle object
    and draws a rectangle.

    arguments:
    t:      Turtle object
    rect:   Rectangle object
    color:  optional color of the Turtle pen.
            Default is black.

    """
    
    if (rect.corner.x, rect.corner.y) != (0, 0):
        t.penup()
        t.fd(rect.corner.x)
        t.lt(90)
        t.fd(rect.corner.y)
        t.rt(90)
        t.pendown()

    for i in range(2):
        t.pencolor(color)
        t.fd(rect.width)
        t.lt(90)
        t.fd(rect.height)
        t.lt(90)


box = Rectangle()
box.width = 100.0
box.height = 200.0
box.corner = Point()
box.corner.x = 0.0
box.corner.y = 0.0

new_box = Rectangle()
new_box.width = 100.0
new_box.height = 200.0
new_box.corner = Point()
new_box.corner.x = 100.0
new_box.corner.y = 100.0

draw_rect(bob, box)

draw_rect(bob, new_box, color = "red")

turtle.mainloop()
```

*Write a function called `draw_circle` that takes a `Turtle` and a `Circle` and draws the `Circle`.*

*__Same notes as above.  Below is this code which can be found in my GitHub repo [here](https://github.com/Sturzgefahr/ThinkPython/blob/master/Think%20Python%20-%20Chapter%2015/draw_circle.py "draw_circle.py").__*

```
import turtle, math
bob = turtle.Turtle()

class Point:
    """Represents a point in 2-D space."""

class Circle:
    """
    Represents a circle.
    
    Attributes: radius, center
    """

    
def draw_circle(t, circle, color = "black"):
    """
    Takes a Circle object and a Turtle object
    and draws a circle.

    arguments:
    t:      Turtle object
    circle: Circle object
    color:  optional color of the Turtle pen.
            Default is black.

    """
    
    t.penup()
    t.fd(circle.center.x)
    t.lt(90)
    t.fd(circle.center.y + circle.radius)
    t.rt(90)
    t.pendown()

    step = (math.pi * 2 * circle.radius)/360

    for i in range(360):
        t.pencolor(color)
        t.fd(step)
        t.rt(1)
    


roundy = Circle()
roundy.radius = 75
roundy.center = Point()
roundy.center.x = 150.0
roundy.center.y = 100.0


draw_circle(bob, roundy, color = "red")

turtle.mainloop()
```